staging/src/k8s.io/cloud-provider/controllers/service/controller.go

```diff
@@ -762,25 +786,62 @@ func (s *Controller) nodeSyncInternal(ctx context.Context, workers int) {
 }
 
 // nodeSyncService syncs the nodes for one load balancer type service
-func (s *Controller) nodeSyncService(svc *v1.Service) bool {
+func (s *Controller) nodeSyncService(svc *v1.Service, oldNodes, newNodes []*v1.Node) bool {
 	if svc == nil || !wantsLoadBalancer(svc) {
 		return false
 	}
 	klog.V(4).Infof("nodeSyncService started for service %s/%s", svc.Namespace, svc.Name)
-	hosts, err := listWithPredicate(s.nodeLister, s.getNodeConditionPredicate())
-	if err != nil {
-		runtime.HandleError(fmt.Errorf("failed to retrieve node list: %v", err))
-		return true
+	newReadyNodes := filterWithPredicate(newNodes, s.getNodeConditionPredicate(true))
+	// For services of ETP=local: we do not want to sync nodes transitioning from NotReady <-> Ready
+	// if these hosts are not actually hosting any endpoints for the service. Doing that leads to
+	// unnecessary updates for LBs which will never send traffic to the nodes in question in any case.
+	if svc.Spec.ExternalTrafficPolicy == v1.ServiceExternalTrafficPolicyTypeLocal {
+		oldReadyNodes := filterWithPredicate(oldNodes, s.getNodeConditionPredicate(true))
+		oldReadyNodeNames := nodeNames(oldReadyNodes)
+		newReadyNodeNames := nodeNames(newReadyNodes)
+		newNotReadyNodes := oldReadyNodeNames.Difference(newReadyNodeNames)
+		newReadyNodes := newReadyNodeNames.Difference(oldReadyNodeNames)
+		// We have four cases to consider here:
+		// a) Node ADD
+		// b) Node DEL
+		// [Node UPDATE]
+		//   c) Ready -> NotReady
+		//   d) NotReady -> Ready
+		// We only want to possibly __not__ sync on c) and d).
+		if len(oldNodes) == len(newNodes) {
+			if newNotReadyNodes.Len() > 0 && !s.serviceHasEndpointsOnNodes(svc, newNotReadyNodes) {
```

Comment:
I think, as written, this will skip sync on services that need sync.  If one node became unready with no endpoints and one became ready (with endpoints) the first case would hit and it would return without evaluating the second case?

Reply:
Yes, quite possibly. I will try to extend my tests with a more exhaustive list of corner cases, one of which should include this.